Remove some redundant / unused binplace directories

ericstj opened the pull request:
I came across these when setting up binplacing for source generators.  Figured I'd clean this up first.

ericstj committed:
Remove some redundant / unused binplace directories

ericstj commented on eng/pipelines/libraries/prepare-for-bin-publish.yml:
```diff
@@ -6,46 +6,24 @@ parameters:
-        targetFolder: $(Build.ArtifactStagingDirectory)/artifacts/bin/testhost
+        sourceFolder: $(Build.SourcesDirectory)/artifacts/bin
+        targetFolder: $(Build.ArtifactStagingDirectory)/artifacts/bin
+        contents: |
```
Consolidated for readability

ericstj commented on eng/pipelines/libraries/prepare-for-bin-publish.yml:
```diff
@@ -6,46 +6,24 @@ parameters:
+      targetFolder: $(Build.ArtifactStagingDirectory)/artifacts/bin
+      contents: |
+        docs/**
+        microsoft.netcore.app.*/**
```
This replaces `pkg` folder and `ref/microsoft.netcore.app` folder

ericstj committed:
Fix runtime pack path

ericstj committed:
Use OuputRid instead of PackageRID

On linux-musl, OutputRID differs from PackageRID.

ViktorHofer commented on Directory.Build.props:
```diff
@@ -208,8 +208,8 @@
 
-    <MicrosoftNetCoreAppRuntimePackDir>$([MSBuild]::NormalizeDirectory('$(ArtifactsBinDir)', 'microsoft.netcore.app.runtime.$(PackageRID)', '$(Configuration)'))</MicrosoftNetCoreAppRuntimePackDir>
-    <MicrosoftNetCoreAppRuntimePackRidDir>$([MSBuild]::NormalizeDirectory('$(MicrosoftNetCoreAppRuntimePackDir)', 'runtimes', '$(PackageRID)'))</MicrosoftNetCoreAppRuntimePackRidDir>
+    <MicrosoftNetCoreAppRuntimePackDir>$([MSBuild]::NormalizeDirectory('$(ArtifactsBinDir)', 'microsoft.netcore.app.runtime.$(OutputRid)', '$(LibrariesConfiguration)'))</MicrosoftNetCoreAppRuntimePackDir>
```
Where is LibrariesConfiguration defined in both a root and a non root build?

ericstj replied:
Here: https://github.com/dotnet/runtime/blob/207a2af755ed370e9c311cbf498ccb9f0a35c972/Directory.Build.props#L98

ericstj replied:
Looks like this is causing failures because this is set very late, after many other properties have derived values from this:
https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/tests/Directory.Build.targets#L268

Any objection to moving that to a props file?

ViktorHofer replied:
Nope, go for it.

ViktorHofer reviewed (commented):
Great change. Wanted to do this myself for quite some time. Thanks a lot.

ericstj commented:
There's still some disagreement between PackageRID / OutputRID values.  I had to use OutputRID to fix the failures where installer didn't agree with libraries for linux-musl builds.   Library builds were producing `linux-musl-*`, but installer was trying to pick up `linux-*`.

After I made the change, installer now fails:
https://dev.azure.com/dnceng/public/_build/results?buildId=1210722&view=logs&j=fa0860f6-441e-50d3-b39c-75f92c2c6627&t=08912796-4720-590b-3e82-a835293fd6c1

This is because libraries is now producing `linux-*` but installer is expecting `linux-musl-*`.  I believe this is because of inconsistent global properties when running libraries vs installer legs (we specify the RID in different ways).  Another annoyance here is that we even have these multiple ways to specify RID, but that's a different problem to fix.  I have to step away from this for some time to get some sleep.

ericstj commented:
This might be the problem.

Installer specifies `OuputRid` only, which makes the `PackageRid` value wrong (which is derived from `RuntimeOS`):
https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/eng/pipelines/installer/jobs/base-job.yml#L216-L219

Libraries only specifies `RuntimeOS` which makes `PackageRid` correct but `OuputRid` wrong (due to `PortableBuild` defaulting to true, whatever that means)
https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/eng/pipelines/libraries/base-job.yml#L67

Some jobs specify both:
https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/eng/pipelines/common/global-build-job.yml#L68

Any recommendations for how to best handle this @safern @ViktorHofer?  We should untangle this by reducing concepts and documenting.

ViktorHofer commented:
@ericstj please see https://github.com/dotnet/runtime/pull/48505#issuecomment-782591528. Unfortunately we had to revert that change because it broke official builds. cc @am11

ericstj committed:
Add RuntimeOS to installer leg for musl

So that we have consistent values of $(PackageRID) for installer and libraries

ericstj commented:
Looking through the yml a bit more, I see we have `_osParameter`.  Should we be using this in more places @safern?  That would centralize at least these properties and avoid the duplication.  For now I made what I felt was the smallest change 🤞

safern reviewed (approved):
LGTM

safern commented:
Yes, we could definitely move this declaration: https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/eng/pipelines/common/global-build-job.yml#L63-L68

Up to `xplat-setup.yml` which is where we set the common/shared properties across all ymls.

ericstj committed:
Move LibrariesConfiguration definition earlier in src/tests

This property was being defined after a number of places where it was
consumed resulting in quite a few inconsistent derived properties.

ericstj commented:
This one is looking like we might have a winner 🤞🤞🤞

ericstj commented:
> This one is looking like we might have a winner 🤞🤞🤞

![image](https://user-images.githubusercontent.com/8918108/123877531-36529880-d8f2-11eb-9b71-85c659f19f11.png)

Jinxed it...  Failure seems non-actionable: https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-54890-merge-d58706d6f6a94bb981/JIT.Methodical.a-dA-D/console.992c771d.log
Will need to rerun.

ericstj commented:
Android failures are https://github.com/dotnet/runtime/issues/42548
```
[04:48:51] fail: Waiting for command timed out: execution may be compromised.
```

Arm64 failure is `profiler.gc` which is occurring on multiple PRs.  The same test passed on an earlier iteration of the same commit hash.  Filed https://github.com/dotnet/runtime/issues/54948

gbalykov commented:
Previously non-oob dlls for libraries were places in `artifacts/bin/pkg/`. Now this dir is deleted and `artifacts/bin/runtime/` contains both non-oob and oob libs. Is there a way to copy only non-oob dlls now?
cc @alpencolt

ViktorHofer commented:
bin/artifacts/microsoft.netcore.app.ref/ or bin/artifacts/microsoft.netcore.app.runtime

gbalykov commented:
@ViktorHofer thanks!
